### PR TITLE
S2259: Support coalescing and conditional access

### LINKED ISSUE #4250: S2259 FP: Lifted operator results in null value in value type comparison
### Repro steps

```cs
public class TokenService
{
    private AccessTokenModel _lastTokenModel;
    public DateTime? LastIssue { get; private set; }

    public string GetAccessToken()
    {
        if (LastIssue + _lastTokenModel?.ExpiresIn > DateTime.Now)
            return _lastTokenModel.Token; // S2259: '_lastTokenModel' is null on at least one execution path.

        ...
    }
}

public class AccessTokenModel
{
    public TimeSpan ExpiresIn { get; set; }
    public string Token { get; set; }
}
```

### Expected behavior

No complaints because: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#lifted-operators

### Actual behavior

> S2259: '_lastTokenModel' is null on at least one execution path.

### Related information

> Microsoft Visual Studio Professional 2019 Version 16.9.4
> Installed Version: Professional
> SonarLint for Visual Studio   4.33.1.28913
.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #4250

Remove suppresion from #6117 after #6812  was implemented.

### CODE REVIEW

File: analyzers/its/expected/Nancy/Nancy--net452-S2259.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+"region":  {
+"startLine":  58,
+"startColumn":  34,
+"endLine":  58,
+"endColumn":  42
+}
+}
+}
+]
+}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/Nancy/Nancy--net452-S2259.json:17]: This is a false positive. `viewName` can only be null if `model` is not null, in which case `viewName.Length` is not accessed.
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/Nancy/src/Nancy/ViewEngines/DefaultViewFactory.cs#L51

File: analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Remote\RemoteActorRefProvider.cs",
+"region":  {
+"startLine":  802,
+"startColumn":  58,
+"endLine":  802,
+"endColumn":  68
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json:15]: False positive. If `deadLetter` is null, the pattern matching will fail.
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Remote/RemoteActorRefProvider.cs#L800

File: analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+"region":  {
+"startLine":  235,
+"startColumn":  56,
+"endLine":  235,
+"endColumn":  71
+}
+}
+}
+]
+}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json:30]: False Positive. `SocketError.ConnectionRefused` is an enum and thus never `null`.
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Remote/Transport/DotNetty/TcpTransport.cs#L233

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Streams\Dsl\Flow.cs",
+"region":  {
+"startLine":  477,
+"startColumn":  76,
+"endLine":  477,
+"endColumn":  81
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:15]: False Positive. `graph` is not checked for null and thus no constraint should be learned.
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Streams/Dsl/Flow.cs#L477

File: analyzers/its/expected/Nancy/Nancy--netstandard2.0-S2259.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+"region":  {
+"startLine":  58,
+"startColumn":  34,
+"endLine":  58,
+"endColumn":  42
+}
+}
+}
+]
+}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/Nancy/Nancy--netstandard2.0-S2259.json:17]: FP, same as above.

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Streams\Dsl\RunnableGraph.cs",
+"region":  {
+"startLine":  163,
+"startColumn":  68,
+"endLine":  163,
+"endColumn":  69
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:28]: FP, similar as above.
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Streams/Dsl/RunnableGraph.cs#L163

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Streams\Dsl\Source.cs",
+"region":  {
+"startLine":  518,
+"startColumn":  65,
+"endLine":  518,
+"endColumn":  71
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:41]: FP, similar as above.
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Streams/Dsl/Source.cs#L518

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Streams\Implementation\Fusing\StreamOfStreams.cs",
+"region":  {
+"startLine":  1217,
+"startColumn":  33,
+"endLine":  1217,
+"endColumn":  39
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:54]: FP
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Streams/Implementation/Fusing/StreamOfStreams.cs#L1217

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Streams\Shape.cs",
+"region":  {
+"startLine":  57,
+"startColumn":  92,
+"endLine":  57,
+"endColumn":  97
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:67]: FP
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Streams/Shape.cs#L57

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+"region":  {
+"startLine":  122,
+"startColumn":  98,
+"endLine":  122,
+"endColumn":  104
+}
+}
+}
+]
+}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:82]: FP
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Streams/Shape.cs#L122

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.cs [hunk 11]
```diff
@@ -15,23 +15,28 @@ public void NullCoalescenceOperator_Null_Noncompliant1()
...
         public void NullCoalesce_Conversion_DownCast(AggregateException arg)
         {
-            var value = arg as Exception ?? new Exception(arg.Message);     // FN Suppressed #6117, arg must be null on the right side
+            var value = arg as Exception ?? new Exception(arg.Message);     // Noncompliant, arg must be null on the right side
         }
 
         public void NullCoalesce_Conversion_UpCast(Exception arg)
         {
-            var value = arg as AggregateException ?? new AggregateException(arg.Message);     // FN Suppressed #6117, arg can be null or another Exception type on the right side
+            var value = arg as AggregateException ?? new AggregateException(arg.Message);     // Noncompliant, arg can be null or another Exception type on the right side
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.cs:34]: I consider this a FP. There is no indication in the code `arg` is ever null.

**Review by Tim-Pohlmann — COMMENTED** [review]
1 true positive to 10 false positives seems like a terrible ratio. Should we really lift the suppression?

**Comment by pavel-mikula-sonarsource** [comment]
Almost all findings here are FPs, and almost all are caused by the same case with `as` operator, not the `??` or `?` logic that was the original motivation for the suppression.

We need to get this on Peach. And eventually fix the `as` in follow up PR.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 12]
```diff
@@ -170,7 +170,7 @@ public void Method2(ReproForIssue2338 obj)
             switch (obj?.Color)
             {
                 case ConsoleColor.Red:
-                    Console.ForegroundColor = obj.Color;  // Compliant, was compliant in the old engine. Should be fixed by MMF-2401
+                    Console.ForegroundColor = obj.Color;  // Noncompliant FP, was compliant in the old engine. Should be fixed by MMF-2401
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:173]: These got fixed in #6993

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 13]
```diff
@@ -1266,6 +1266,53 @@ public void StringIsNullOrEmpty_Property2(string arg)
...
+        {
+            var first = message as IFirst;
+            var second = message as ISecond;
+            if(first != null)
+            {
+                // Do something
+            }
+            else if(second?.Property is string)
+            {
+                var str = second.Property; // Noncompliant FP
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:1309]: This is fixed in #7021

**Pull request merged**